# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import pandas as pd
import numpy as np


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather.csv` file. 

#### First import it into a data frame called `austin`.

In [49]:
# Your code here
df = pd.read_csv('C:/Users/Ironhack/Documents/Ironhack/Week1/data-labs/module-3/lab-intro-to-ml/austin_weather.csv')
df.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [3]:
# Your code here
df.dtypes

Date                          object
TempHighF                      int64
TempAvgF                       int64
TempLowF                       int64
DewPointHighF                 object
DewPointAvgF                  object
DewPointLowF                  object
HumidityHighPercent           object
HumidityAvgPercent            object
HumidityLowPercent            object
SeaLevelPressureHighInches    object
SeaLevelPressureAvgInches     object
SeaLevelPressureLowInches     object
VisibilityHighMiles           object
VisibilityAvgMiles            object
VisibilityLowMiles            object
WindHighMPH                   object
WindAvgMPH                    object
WindGustMPH                   object
PrecipitationSumInches        object
Events                        object
dtype: object

In [4]:
# Your code here
df.describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [5]:
# Your code here
df.head(5)

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [6]:
# Your code here
df['DewPointAvgF'].unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [ ]:
# Your observation here
#Because the '' (string)

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [7]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [8]:
# Your code here
df[wrong_type_columns] = df[wrong_type_columns].apply(pd.to_numeric, errors='coerce') 
df

2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
5             33.0                 79.0                63.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
8             28.0                 92.0                76.0   
9             21.0                 75.0                60.0   
10            23.0                 76.0                54.0   
11            30.0                 83.0                68.0   
12            19.0                 83.0                55.0   
13            18.0                 75.0                49.0   
14            38.0                 89.0                68.0   
15            15.0                 87.0                59.0   
16             2.0                 50.0                38.0   
17             4.0                 68.0                43.0   
18            30.0                 93.0                75.0   
19            50.0                 93.0                80.0   
20            54.0                 93.0                81.0   
21            20.0                 93.0                53.0   
22            33.0                 84.0                68.0   
23            17.0                 87.0                51.0   
24            20.0                 70.0                41.0   
25            16.0                 49.0                35.0   
26            23.0                 55.0                38.0   
27            15.0                 45.0                31.0   
28            23.0                 59.0                45.0   
29            18.0                 64.0                39.0   
...            ...                  ...                 ...   
1289          66.0                 91.0                63.0   
1290          67.0                 88.0                62.0   
1291          66.0                 85.0                60.0   
1292          60.0                 85.0                58.0   
1293          63.0                 88.0                60.0   
1294          65.0                 88.0                61.0   
1295          62.0                 87.0                59.0   
1296          60.0                 82.0                56.0   
1297          58.0                 87.0                57.0   
1298          58.0                 88.0                58.0   
1299          59.0                 88.0                59.0   
1300          63.0                 88.0                59.0   
1301          61.0                 88.0                58.0   
1302          63.0                100.0                65.0   
1303          65.0                100.0                67.0   
1304          66.0                 88.0                63.0   
1305          65.0                 88.0                61.0   
1306          63.0                 88.0                60.0   
1307          61.0                 85.0                56.0   
1308          61.0                 85.0                56.0   
1309          61.0                 88.0                58.0   
1310          66.0                 85.0                58.0   
1311          56.0                 91.0                60.0   
1312          63.0                 91.0                61.0   
1313          61.0                 91.0                58.0   
1314          61.0                 82.0                54.0   
1315          55.0             

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [9]:
# Your code here
df.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [13]:
# Your code here
missing_values =  df[df.isnull().any(axis=1)]
missing_values 

8  2017-07-01         98        89        79           76.0          72.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1308  2017-07-21        104        91        77           74.0          68.0   
1313  2017-07-26        103        91        78           75.0          70.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
4             36.0                 86.0                71.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
42            41.0                 81.0                71.0   
51            30.0                 82.0                74.0   
66            47.0                 93.0                85.0   
95            30.0                 78.0                57.0   
102           63.0                 97.0                78.0   
103           66.0                 90.0                74.0   
104           28.0                 93.0                58.0   
105           33.0                 60.0                48.0   
113           61.0                 90.0                79.0   
126           62.0                 90.0                73.0   
139           58.0                 87.0                69.0   
152           61.0                 90.0                70.0   
154           66.0                 97.0                79.0   
155           63.0                 84.0                66.0   
164           68.0                 82.0                67.0   
169           64.0                 87.0                67.0   
174            NaN                 95.0                75.0   
175            NaN                 95.0                77.0   
176            NaN                  NaN                 NaN   
177            NaN                 94.0                72.0   
182           68.0                 94.0                72.0   
183           68.0                 85.0                70.0   
188           70.0                 88.0                69.0   
189           71.0                 85.0                69.0   
248           62.0                 71.0                52.0   
278           57.0                 90.0                74.0   
280           64.0                 93.0                78.0   
...            ...                  ...                 ...   
1101          67.0                 93.0                74.0   
1102          47.0                 97.0                72.0   
1105          18.0                 48.0                38.0   
1106          36.0                 90.0                70.0   
1107          39.0                 97.0                74.0   
1125          50.0                100.0                79.0   
1134          21.0                 82.0                52.0   
1140          31.0                 61.0                57.0   
1142          51.0                100.0                89.0   
1143          63.0                 93.0                76.0   
1150          42.0                 84.0                63.0   
1163          33.0                 93.0                77.0   
1164          64.0                 93.0                76.0   
1171          64.0                 93.0                76.0   
1172          34.0                 93.0                61.0   
1173          33.0                 78.0                56.0   
1174          59.0                 93.0                79.0   
1175          59.0                 93.0                70.0   
1209          58.0                100.0                77.0   
1213          57.0                 90.0                80.0   
1215          64.0                 87.0                71.0   
1229          47.0                 93.0                66.0   
1236          66.0                 90.0                76.0   
1244          65.0                 94.0                72.0   
1259          66.0                 93.0                70.0   
1269          62.0                 87.0                62.0   
1288          66.0          

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [14]:
# Your code here
missing_values.shape[0]
df.shape[0]

1319

#### Calculate the ratio of missing rows to total rows

In [15]:
# Your code here
missing_values.shape[0]/df.shape[0]*100

10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [16]:
# Your code here
df.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [17]:
# Your code here
df['PrecipitationSumInches'].isna().sum()/missing_values.shape[0] * 100

91.17647058823529

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [18]:
# Your code here 
df.drop(columns='PrecipitationSumInches', inplace=True, axis = 1)

# Print `austin` to confirm the column is indeed removed

df

103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
5             33.0                 79.0                63.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
8             28.0                 92.0                76.0   
9             21.0                 75.0                60.0   
10            23.0                 76.0                54.0   
11            30.0                 83.0                68.0   
12            19.0                 83.0                55.0   
13            18.0                 75.0                49.0   
14            38.0                 89.0                68.0   
15            15.0                 87.0                59.0   
16             2.0                 50.0                38.0   
17             4.0                 68.0                43.0   
18            30.0                 93.0                75.0   
19            50.0                 93.0                80.0   
20            54.0                 93.0                81.0   
21            20.0                 93.0                53.0   
22            33.0                 84.0                68.0   
23            17.0                 87.0                51.0   
24            20.0                 70.0                41.0   
25            16.0                 49.0                35.0   
26            23.0                 55.0                38.0   
27            15.0                 45.0                31.0   
28            23.0                 59.0                45.0   
29            18.0                 64.0                39.0   
...            ...                  ...                 ...   
1289          66.0                 91.0                63.0   
1290          67.0                 88.0                62.0   
1291          66.0                 85.0                60.0   
1292          60.0                 85.0                58.0   
1293          63.0                 88.0                60.0   
1294          65.0                 88.0                61.0   
1295          62.0                 87.0                59.0   
1296          60.0                 82.0                56.0   
1297          58.0                 87.0                57.0   
1298          58.0                 88.0                58.0   
1299          59.0                 88.0                59.0   
1300          63.0                 88.0                59.0   
1301          61.0                 88.0                58.0   
1302          63.0                100.0                65.0   
1303          65.0                100.0                67.0   
1304          66.0                 88.0                63.0   
1305          65.0                 88.0                61.0   
1306          63.0                 88.0                60.0   
1307          61.0                 85.0                56.0   
1308          61.0                 85.0                56.0   
1309          61.0                 88.0                58.0   
1310          66.0                 85.0                58.0   
1311          56.0                 91.0                60.0   
1312          63.0                 91.0                61.0   
1313          61.0                 91.0                58.0   
1314          61.0              

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [50]:
# Your code here
austin_fixed = df.interpolate(inplace = False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [51]:
# Your code here
austin_fixed.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
PrecipitationSumInches        0
Events                        0
dtype: int64

In [22]:
df.isna().sum()

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [23]:
# Your code here:
austin_fixed['Events'].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [ ]:
# Your answer:
# 3 events: Fog , Rain , Thunderstorm

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [52]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here

for event in event_list:
    austin_fixed[event] = 0

# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0     2013-12-21         74        60        45            67           49   
1     2013-12-22         56        48        39            43           36   
2     2013-12-23         58        45        32            31           27   
3     2013-12-24         61        46        31            36           28   
4     2013-12-25         58        50        41            44           40   
5     2013-12-26         57        48        39            39           36   
6     2013-12-27         60        53        45            41           39   
7     2013-12-28         62        51        40            43           39   
8     2013-12-29         64        50        36            49           41   
9     2013-12-30         44        40        35            31           26   
10    2013-12-31         55        46        36            31           28   
11    2014-01-01         69        54        39            51           42   
12    2014-01-02         55        44        33            39           26   
13    2014-01-03         58        43        28            37           22   
14    2014-01-04         71        57        42            55           48   
15    2014-01-05         59        47        34            54           32   
16    2014-01-06         36        29        22            15            8   
17    2014-01-07         48        35        22            29           11   
18    2014-01-08         53        47        40            51           45   
19    2014-01-09         70        62        53            60           55   
20    2014-01-10         72        65        57            64           61   
21    2014-01-11         75        62        48            54           37   
22    2014-01-12         67        57        46            58           47   
23    2014-01-13         71        57        43            60           36   
24    2014-01-14         76        57        37            31           25   
25    2014-01-15         64        52        40            29           23   
26    2014-01-16         72        56        40            31           27   
27    2014-01-17         64        54        43            31           20   
28    2014-01-18         70        54        37            41           33   
29    2014-01-19         72        59        45            41           30   
...          ...        ...       ...       ...           ...          ...   
1289  2017-07-02        100        89        78            75           72   
1290  2017-07-03         99        88        76            74           71   
1291  2017-07-04        101        90        78            74           72   
1292  2017-07-05         99        89        78            74           70   
1293  2017-07-06        100        89        77            74           71   
1294  2017-07-07         99        87        75            74           71   
1295  2017-07-08        100        88        75            71           67   
1296  2017-07-09        101        89        77            71           67   
1297  2017-07-10        100        88        76            72           65   
1298  2017-07-11        101        89        77            73           69   
1299  2017-07-12        101        89        77            73           70   
1300  2017-07-13        102        90        77            74           70   
1301  2017-07-14        101        89        77            74           69   
1302  2017-07-15        103        87        71            73           70   
1303  2017-07-16         99        84        68            71           69   
1304  2017-07-17         98        88        77            75           71   
1305  2017-07-18        100        88        76            74           70   
1306  2017-07-19        101        89        77            74           70   
1307  2017-07-20        103        90        77            75           69   
1308  2017-07-21        104       

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [53]:
# Your code here
austin_fixed['Snow'] = austin_fixed['Events'].str.contains('Snow').astype(int)
austin_fixed['Fog'] = austin_fixed['Events'].str.contains('Fog').astype(int)
austin_fixed['Rain'] = austin_fixed['Events'].str.contains('Rain').astype(int)
austin_fixed['Thunderstorm'] = austin_fixed['Events'].str.contains('Thunderstorm').astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [54]:
# Your code here
austin_fixed[['Events','Snow', 'Fog', 'Rain', 'Thunderstorm']]

Events  Snow  Fog  Rain  Thunderstorm
0     Rain , Thunderstorm     0    0     1             1
1                             0    0     0             0
2                             0    0     0             0
3                             0    0     0             0
4                             0    0     0             0
5                             0    0     0             0
6                             0    0     0             0
7                             0    0     0             0
8                             0    0     0             0
9                             0    0     0             0
10                            0    0     0             0
11                            0    0     0             0
12                            0    0     0             0
13                            0    0     0             0
14                            0    0     0             0
15                            0    0     0             0
16                            0    0     0             0
17                            0    0     0             0
18                   Rain     0    0     1             0
19                    Fog     0    1     0             0
20                   Rain     0    0     1             0
21                            0    0     0             0
22                   Rain     0    0     1             0
23                            0    0     0             0
24                            0    0     0             0
25                            0    0     0             0
26                            0    0     0             0
27                            0    0     0             0
28                            0    0     0             0
29                            0    0     0             0
...                   ...   ...  ...   ...           ...
1289                          0    0     0             0
1290                          0    0     0             0
1291                          0    0     0             0
1292                          0    0     0             0
1293         Thunderstorm     0    0     0             1
1294  Rain , Thunderstorm     0    0     1             1
1295         Thunderstorm     0    0     0             1
1296                          0    0     0             0
1297                          0    0     0             0
1298                          0    0     0             0
1299                          0    0     0             0
1300                          0    0     0             0
1301                          0    0     0             0
1302  Rain , Thunderstorm     0    0     1             1
1303                          0    0     0             0
1304  Rain , Thunderstorm     0    0     1             1
1305                          0    0     0             0
1306                          0    0     0             0
1307                          0    0     0             0
1308                          0    0     0             0
1309                          0    0     0             0
1310  Rain , Thunderstorm     0    0     1             1
1311  Rain , Thunderstorm     0    0     1             1
1312                          0    0     0             0
1313                          0    0     0             0
1314                          0    0     0             0
1315                          0    0     0             0
1316                          0    0     0             0
1317                          0    0     0             0
1318                          0    0     0             0

[1319 rows x 5 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [62]:
# Your code here
austin_fixed.drop(['Events'], inplace=True, axis=1)
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  735223         74        60        45            67           49   
1  735224         56        48        39            43           36   
2  735225         58        45        32            31           27   
3  735226         61        46        31            36           28   
4  735227         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches Snow Fog Rain Thunderstorm  
0                   0.46    0   0    1            1  
1                      0    0   0    0            0  
2                      0    0   0    0            0  
3                      0    0   0    0            0  
4                      T    0   0    0            0  

[5 rows x 24 columns]

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [58]:
# Your code here
from datetime import datetime as dt
austin_fixed['Date'] = austin_fixed['Date'].apply(lambda x: dt.strptime(x, '%Y-%m-%d'))

In [60]:
austin_fixed['Date'] = austin_fixed['Date'].apply(lambda x: x.toordinal())

#### Print `austin_fixed` to check your `Date` column.

In [61]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  735223         74        60        45            67           49   
1  735224         56        48        39            43           36   
2  735225         58        45        32            31           27   
3  735226         61        46        31            36           28   
4  735227         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events Snow Fog Rain Thunderstorm  
0                   0.46  Rain , Thunderstorm    0   0    1            1  
1                      0                         0   0    0            0  
2                      0                         0   0    0            0  
3                      0                         0   0    0            0  
4                      T                         0   0    0            0  

[5 rows x 25 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [69]:
# Your code here:
X = austin_fixed[['TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']]

y = austin_fixed['TempAvgF']

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [68]:
#Your code here:
from sklearn.model_selection import train_test_split 

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [71]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(
    X, #features
    y, #target
    test_size = .2
)

X_train.head()

TempHighF  TempLowF DewPointHighF DewPointAvgF DewPointLowF  \
668         87        58            68           58           52   
118         78        59            61           57           51   
127         96        73            70           66           41   
765         75        56            57           51           37   
919         96        75            75           71           68   

    HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
668                  86                 61                 35   
118                  93                 68                 43   
127                  87                 51                 15   
765                  93                 66                 38   
919                  87                 65                 42   

    SeaLevelPressureHighInches SeaLevelPressureAvgInches  ...  \
668                      30.17                     30.08  ...   
118                      30.23                     30.17  ...   
127                      29.63                     29.55  ...   
765                      30.06                     29.94  ...   
919                      30.14                     30.06  ...   

    VisibilityHighMiles VisibilityAvgMiles VisibilityLowMiles WindHighMPH  \
668                  10                 10                 10          13   
118                  10                  6                  2           9   
127                  10                  6                  3          14   
765                  10                 10                 10          10   
919                  10                 10                 10           8   

    WindAvgMPH WindGustMPH Snow  Fog  Rain  Thunderstorm  
668          4          20    0    0     0             0  
118          2          16    0    0     1             0  
127          6          24    0    0     0             0  
765          4          20    0    0     0             0  
919          2          15    0    0     0             0  

[5 rows x 21 columns]

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:


# Bonus Challenge 2

As explained in the README, the main purpose of this lab is to show you the typical process of preparing data for machine learning which sometimes takes up 90% of your time. Data cleaning is a valuable skill to learn and you need to be proficient at various techniques including the ones we showed you above as well as others you'll learn in the future. In the real world this skill will help you perform your job successfully and efficiently.

Now that we're done with data praparation, if you want to expeirence what you'll do in the rest 10% of your time, let's make the final leap.

We will use scikit-learn's [*Support Vector Machines*](https://scikit-learn.org/stable/modules/svm.html) to compute the fit of our training data set, the test on our test data set.

#### In the cell below, import `svm` from `sklearn`:

In [ ]:
# Your code here


#### Now, call `svm.SVC.fit()` on `X_train` and `y_train`. Assign the returned value to a variable called `clf` which stands for *classifier*. Then obtain the test score for `X_test` and `y_test` by calling `clf.score()`.

In [ ]:
# Your code here


#### You now see the model fit score of your test data set. If it's extremely low, it means your selected model is not a good fit and you should try other models.

#### In addition to fitting `X_train`, `y_train`, `X_test`, and `y_test`, you can also fit `X_ts_train`, `y_ts_train`, `X_ts_test`, and `y_ts_test` if you completed Bonus Challenge 1.

In [ ]:
# Your code here


#### We hope you have learned a lot of useful stuff in this lab!